src/raftstore/store/store.rs

```diff
@@ -1276,15 +1293,29 @@ impl<T: Transport, C: PdClient> Store<T, C> {
 
         let res = peer::check_epoch(peer.region(), msg);
         if let Err(Error::StaleEpoch(msg, mut new_regions)) = res {
-            // Attach the next region which might be split from the current region. But it doesn't
-            // matter if the next region is not split from the current region. If the region meta
+            // Attach the region which might be split from the current region. But it doesn't
+            // matter if the region is not split from the current region. If the region meta
             // received by the TiKV driver is newer than the meta cached in the driver, the meta is
             // updated.
-            if let Some((_, &next_region_id)) = self.region_ranges
+            let new_region_id = if self.cfg.right_derive_when_split {
```

Comment:
can we extract the logic and check individually?  I am concerned the correctness here.